### ISSUE TITLE
Ansible installed globally confuse binary installed in a python virtualenv (due to PYTHONPATH)

### ISSUE THREAD

**Issue by allan-simon** [ansible/ansible#16816]
##### ISSUE TYPE
Bug Report

##### COMPONENT NAME
ansiballz/ziploader for modules

##### ANSIBLE VERSION
```
ansible 2.1.0.0
config file =
configured module search path = Default w/o overrides
```

##### CONFIGURATION
Nothing special

##### SUMMARY
Ansible is installed globally on my developer machine
I use it to provision the same machine
One of these tasks is 
``` yml
    - name: migrate database
      shell: "/tm/my_virtual_env/bin/alembic upgrade head"
```
This command fails with an ImportError when run in ansible, but it does when run manually
After investigation, Ansible set one more environment variable :  `PYTHONPATH=` (empty value)
which seems to confuse the python script `alembic`
if I do
``` yml
    - name: migrate database
      shell: "env -i /tm/my_virtual_env/bin/alembic upgrade head"
```
Then it works

##### EXPECTED RESULTS
Ansible should not mess around with environment variable, it should be transparent
```
```

**Comment by bcoca** [comment]
Ansible needs to mess with the python path to be able to load it's modules (those that use built in helpers) and execute them, still it should not interfere with other tasks.

**Comment by allan-simon** [comment]
yes, sorry if I was rude with "mess", I only meant for subprocess , it's perfectly fine as you said if it does that for internal tasks

**Comment by abadger** [comment]
@allan-simon I finally had a chance to look at this today and I am unable to reproduce.  I'll need more information from you about how to create a virtualenv to work against, how I might need to configure alembic, what the error messages are, and what is shown when it works.  Here's what I've tried so far:

```
$ cat /srv/ansible/troubleshooting/16816.yml              *[b599477]  (20:23:05)
---
- hosts: localhost
  gather_facts: False
  tasks:
  - pip:
      virtualenv: /var/tmp/myvenv
      name: alembic
      state: present
  - copy:
      content: |
        [alembic]
        script_location=/var/tmp/myvenv
      dest: /var/tmp/myvenv/alembic.ini
  - copy:
      content: "\n"
      dest: /var/tmp/myvenv/env.py
  - shell:  "/var/tmp/myvenv/bin/alembic upgrade head chdir=/var/tmp/myvenv"
```

```
[pts/6@roan /srv/ansible/stable-2.1]$ ansible-playbook --version                              *[b599477]  (20:25:20)
ansible-playbook 2.1.0.0 (detached HEAD b599477242) last updated 2016/09/21 20:00:52 (GMT -700)
  lib/ansible/modules/core: (stable-2.1 2b9300ec28) last updated 2016/09/15 09:43:53 (GMT -700)
  lib/ansible/modules/extras: (stable-2.1 b9853e9e8f) last updated 2016/09/15 09:44:01 (GMT -700)
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
[pts/6@roan /srv/ansible/stable-2.1]$ ansible-playbook /srv/ansible/troubleshooting/16816.yml
 [WARNING]: provided hosts list is empty, only localhost is available


PLAY [localhost] ***************************************************************

TASK [pip] *********************************************************************
ok: [localhost]

TASK [copy] ********************************************************************
ok: [localhost]

TASK [copy] ********************************************************************
ok: [localhost]

TASK [command] *****************************************************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=4    changed=1    unreachable=0    failed=0
```

Thanks.

needs_info

**Comment by jctanner** [comment]
@allan-simon I would also be interested to see the result of this task ...

```
    - shell: env | fgrep -i python
      failed_when: False
```

needs_info

**Comment by allan-simon** [comment]
hello when I arrive in monday at the office I will try to reproduce the bug and put here all the requested information

**Comment by abadger** [comment]
@allan-simon Were you able to reproduce and get more information?

needs_info

**Comment by ansibot** [comment]
@allan-simon ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@allan-simon ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@allan-simon ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jflbr** [comment]
I reproduced the "bug" when I was trying to execute a python script in a virtualenv from an ansible task.

I fixed it by using the absolute python interpreter path of the virtualenv

```
    - name: migrate database
      shell: "{{ansible_env.VIRTUAL_ENV_PATH }}/bin/python backend/migration/migrate_db.py"
      args:...
```

**Comment by abadger** [comment]
@jflbr Do you have a series of steps for creating the virtualenv with the migrate_db script and then trying to execute it?  I was stymied by not being able to reproduce the problem.

**Comment by jflbr** [comment]
I don't create the virtualenv with `migrate_db.py` file. I create it in a previous ansible task as follows:

```  
      - name: creating virtual environment
      shell: >
        tox -r -e {{ ansible_env.ENV_NAME }}
      args:
        chdir: "{{ ansible_env.APP_PATH }}"
        creates: "{{ ansible_env.VIRTUAL_ENV_PATH }}"
```
Next, I execute a db migration task as shown in my previous post:
```
    - name: migrate database
      shell: "{{ansible_env.VIRTUAL_ENV_PATH }}/bin/python backend/migration/migrate_db.py"
      args:
        ...
```

The bug appeared to me when I tried to execute `backend/migration/migrate_db.py` script without specifying the python interpreter which is inside of the virtualenv. I had to do it with the absolute path `{{ansible_env.VIRTUAL_ENV_PATH }}/bin/python`.

**Comment by abadger** [comment]
@jflbr, Was he migrate_db.py script createdd by the python inside of the virtualenv?  What is its shebang line?

**Comment by abadger** [comment]
needs_info

**Comment by ansibot** [comment]
@allan-simon Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information. 

Here are the items we could not find in your description:
- component name

Please set the description of this issue with this template:
https://raw.githubusercontent.com/ansible/ansible/devel/.github/ISSUE_TEMPLATE.md

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by abadger** [comment]
Closing this bug for lack of information from the reporters.  If you are suffering from similar symptoms, please open a new issue with details on how to reproduce the bug so that we can figure out what the code is doing when this bug is exposed.

thanks!
